src/engines/TimeToLive.js

```diff
@@ -0,0 +1,149 @@
+import DefaultHashTable from '../dataStructure/HashTable';
+import DoublyLinkedList from '../dataStructure/DoublyLinkedList';
+import * as hashTableProp from '../hashTableSymbol';
+
+function TimeToLive({ HashTable = DefaultHashTable, defaultTTL } = {}) {
+  const store = new HashTable();
+  const timePartition = new HashTable();
+  const timeIndexInterval = 5 * 60 * 1000; // milliseconds.
+
+  let lowestTimePartition = Date.now();
+
+  this.add = (key, value, ttl = defaultTTL) => {
+    if (!ttl || !Number.isInteger(ttl) || ttl <= 0)
+      throw Error(
+        'Expected ttl value (should be positive integer). ' +
+          'you can have to mention it in add method or mention as defaultTTL at constructor',
+      );
+
+    const expireTTL = Date.now() + ttl;
+    const bucket = getTimeBucket(expireTTL);
+    bucket.addFirst(key);
+    const tNode = bucket.getFirstNode();
+    const payload = { value, ttl: expireTTL, tNode };
+    store[hashTableProp.add](key, payload);
+  };
+
+  this.get = key => {
+    const payload = store[hashTableProp.get](key);
+    if (payload) {
+      const { ttl, value } = payload;
+      if (checkIfElementExpire({ ttl, key })) return undefined;
+      else return value;
+    }
+    return undefined;
+  };
+
+  this.has = key => {
+    return (
+      store[hashTableProp.has](key) &&
+      !checkIfElementExpire({ ttl: store[hashTableProp.get](key), key })
+    );
+  };
+
+  this.remove = key => {
+    if (this.has(key)) {
+      const { ttl, tNode } = store[hashTableProp.get](key);
+      const timeBucket = getTimeBucket(ttl);
+      timeBucket.remove(tNode);
+      store[hashTableProp.remove](key);
+    }
+  };
+
+  this.size = () => {
+    return store[hashTableProp.size]();
+  };
+
+  this.runGC = () => {
+    const cleanTo = getBackwardTimeIndex({ time: Date.now(), interval: timeIndexInterval });
```

Comment:
Just realize maybe not for this version but for future version
a) Providing interval all the time is not very functional programmy. We need more like a createIntervalHandler function that gets the defaultTTL and calculates the best index and then returns an object with methods getPreviousTimeIndex and getCurrentTimeIndex based only on time input
b) The whole timepartition thing is kind of a data structure ;-)